### PR TITLE
8255936: "parsing found no loops but there are some" assertion failure with Shenandoah

### PR DESCRIPTION
**Pull request by rwestrel**
This is a Shenandoah bug but the proposed fix is in shared code.

In an infinite loop, a barrier is located right after the loop head
and above the never branch. When the barrier is expanded, control flow
is added between the loop and the never branch. During loop
verification the assert fires because it doesn't expect any control
flow between the never branch and the loop head.

While it would have been nice to fix this Shenandoah issue in
Shenandoah code, I think the cleaner fix is to preserve the invariant
that the never branch is always right after the loop head in an
infinite loop. In the proposed patch, this is achieved by moving all
uses of the loop head to the never branch when it's constructed.

/cc shenandoah,hotspot-compiler

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Testing

|     | Linux aarch64 | Linux arm | Linux ppc64le | Linux s390x | Linux x64 | Linux x86 | Windows x64 | macOS x64 |
| --- | ----- | ----- | ----- | ----- | ----- | ----- | ----- | ----- |
| Build | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (1/1 passed) | ✔️ (6/6 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) | ✔️ (2/2 passed) |
| Test (tier1) |    |     |     |     |  ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) | ✔️ (9/9 passed) |

### Issue
 * [JDK-8255936](https://bugs.openjdk.java.net/browse/JDK-8255936): "parsing found no loops but there are some" assertion failure with Shenandoah

### Reviewers
 * [Tobias Hartmann](https://openjdk.java.net/census#thartmann) (@TobiHartmann - **Reviewer**)
 * [Roman Kennke](https://openjdk.java.net/census#rkennke) (@rkennke - **Reviewer**)

### Download
`$ git fetch https://git.openjdk.java.net/jdk pull/1073/head:pull/1073`
`$ git checkout pull/1073`

### CODE REVIEW

**Comment by rwestrel** [comment]
I rebased the change

File: src/hotspot/share/opto/loopnode.cpp [hunk 1]
```diff
@@ -4569,14 +4569,13 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
           Node* cfg = NULL;       // Find the One True Control User of m
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:4569]: Should be removed.

File: src/hotspot/share/opto/loopnode.cpp [hunk 2]
```diff
@@ -4569,14 +4569,13 @@ int PhaseIdealLoop::build_loop_tree_impl( Node *n, int pre_order ) {
...
           for (DUIterator_Fast jmax, j = m->fast_outs(jmax); j < jmax; j++) {
             Node* x = m->fast_out(j);
-            if (x->is_CFG() && x != m && x != iff)
-              { cfg = x; break; }
+            if (x != m && x != iff && !x->is_Phi()) {
+                _igvn.rehash_node_delayed(x);
+                int nb = x->replace_edge(m, if_t);
+                assert(nb > 0, "use should have been updated");
+              --j, jmax -= nb;
```
Comment (TobiHartmann) [inline src/hotspot/share/opto/loopnode.cpp:4576]: Indentation is wrong.

**Review by TobiHartmann — APPROVED** [review]
Otherwise looks good to me.

File: src/hotspot/share/opto/loopnode.cpp (hunk 2 above)
Comment (rwestrel) [reply src/hotspot/share/opto/loopnode.cpp:4575]: Updated. Thanks for the review.

**Comment by TobiHartmann** [comment]
@rwestrel please don't use force-push because it prevents getting incremental changes in the PR ("New changes since you last viewed").

**Review by TobiHartmann — APPROVED** [review]

**Comment by navyxliu** [comment]
hi, @rwestrel  

Can we have an assertion to make  this loop invariance more prominent? 
`I think the cleaner fix is to preserve the invariant that the never branch is always right after the loop head in an infinite loop. `

**Comment by rwestrel** [comment]
> Can we have an assertion to make this loop invariant more prominent?
> `I think the cleaner fix is to preserve the invariant that the never branch is always right after the loop head in an infinite loop. `

Thanks for the suggestion. Actually I reworked the fix so it has nothing to do with the previous one.

**Comment by rwestrel** [comment]
I pushed a new fix that's different from the previous one. I realized that NeverBranchNode::Ideal() should remove the NeverBranch once the barrier is expanded but we don't run igvn after barrier expansion which is a mistake.

**Comment by navyxliu** [comment]
yes, looks good to me. 
Happy to see a more elegant way to solve this problem.

**Review by TobiHartmann — APPROVED** [review]
The reworked fix looks good to me too.

**Review by rkennke — APPROVED** [review]
Looks good to me! Thanks!

**Pull request closed without merging**